src/librustc/hir/mod.rs

```diff
@@ -145,7 +145,27 @@ pub struct Lifetime {
     /// HIR lowering inserts these placeholders in type paths that
     /// refer to type definitions needing lifetime parameters,
     /// `&T` and `&mut T`, and trait objects without `... + 'a`.
-    pub name: Name,
+    pub name: LifetimeName,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Copy)]
+pub enum LifetimeName {
+    Elided,
+    Underscore,
+    Static,
+    Name(Name),
+}
```

Comment:
@cramertj 
Actually, can you tweak the naming slightly.
```rust
enum LifetimeKind { // This is not a name
    Implicit, // `Elided` is not correct, because implicit can be both elided and inferred, but we don't know what it is when lowering into HIR
    Underscore,
    Static,
    Named(Name), // Named lifetime
}
```

Reply:
(The variants are going to change anyway when lifetimes are first resolved, then lowered into HIR.)

Reply:
I think what we want is name resolution to give us a `Def`, and the existing `resolve_lifetimes` pass to do elision and compute the semantic representation from `Def::Lifetime(DefId)`.

Reply:
@petrochenkov I don't object to calling "Implicit" for sure, but I'm curious: what does "elided" mean to you? To me, it means "not written". I would then say that the behavior around elided lifetimes depends on context: in function signatures, we use defaulting rules (as specified by the lifetime elision RFCs plus the object-lifetime RFCs), and in function bodies, we use inference.

Reply:
Using "elided" to mean literally "not written" may be correct wrt the English language but then "elision rules" doesn't mean much anymore and we have to say "default lifetime rules" I guess? And "lifetime defaulting" for the process of picking a lifetime.

Reply:
So personally, when I explain to people, I say something like "sometimes you can elide lifetimes. The meaning of this depends on context. In a function signature, we use the following system of defaults. etc". In other words, I think I do use the terms the way you described. =) But also I rely on context. That is, I might say "elision rules" when it's clear we're talking about function signatures, in which case it just means "defaulting rules" (but including, naturally, the object defaults).

In any case, it seems clear that calling things elided is ripe for confusion. I prefer either "implicit", "not written", or "not present", all of which seem reasonably clear.

Reply:
@nikomatsakis 
Well, I usually associate "elided" with "lifetime elision RFCs", i.e. "defaulting", not inference.
It's useful to have some separate words for "not-written-in-signature" and "not-written-in-body", and "elided" and "inferred" already kinda serve this purpose given that "lifetime elision" is a commonly used to describe what happens in signatures.

Reply:
That makes some sense. I guess I prefer to have words that match their English meanings to the extent possible, though, and to me "elided" just means "not written" (whereas "defaulted" seems more specific -- we assigned it a default value). I admit that the distinction between *defaulted* (based on some simple rules) and *inferred* (based on complex analysis) is a subtle one as well. =)

Reply:
Regardless, the question at hand is whether to use elided *here* with the *broader* meaning, right? In which case, "Implicit" or "Not written" seems fine.